torch/autograd/graph.py

```diff
@@ -439,27 +440,60 @@ def inner_hook(grad: torch.Tensor):
 
         return inner_hook
 
-    class Handle(RemovableHandle):
-        handles: Tuple[RemovableHandle, ...]
+    handles: List[RemovableHandle] = []
+    for i, t in enumerate(tensors):
+        handles.append(t.register_hook(get_inner_hook(i)))
 
-        def __init__(self, handles: Tuple[RemovableHandle, ...]):
-            self.handles = handles
+    return _MultiHandle(tuple(handles))
 
-        def remove(self):
-            for handle in self.handles:
-                handle.remove()
 
-        def __getstate__(self):
-            return self.handles
+def _register_any_hook(
+    tensors: Sequence[torch.Tensor],
+    fn: Callable[[Any], None],
+):
+    """
+    Register an "any" backward hook.
 
-        def __setstate__(self, state):
-            self.handles = state
+    The hook is called exactly once when the gradient with respect to *any* of
+    the tensors in :attr:`tensors` is computed and does not run thereafter. It
+    is a wrapper over :meth:`torch.Tensor.register_hook` that ensure that only
+    the first hook that fires actually runs.
 
-    handles: List[RemovableHandle] = []
-    for i, t in enumerate(tensors):
-        handles.append(t.register_hook(get_inner_hook(i)))
+    The hook should not modify its arguments.
+
+    This function returns a handle with a method ``handle.remove()`` that removes the hook.
+    """
+    ran_hook = False
+
+    @functools.wraps(fn)
+    def wrapped_fn(*args, **kwargs):
+        nonlocal ran_hook
+        if not ran_hook:
+            ran_hook = True
```

Comment:
IIUC, this is not thread-safe, as we would need to ensure atomicity of the read and write to `ran_hook`. For the FSDP use-case, we do not need thread safety since we do not run backward with multiple threads. Should we make this thread-safe, or should we just document that it is not thread safe?

Reply:
Maybe do something like the following?

```python
# outside hook
lock = threading.lock()
ran_hook = default_dict(bool)

# inside hook
id = torch._C._current_graph_task_id()
assert id != -1, "Expect hook to be executed during backward"

lock.acquire()
prev, ran_hook[id] = ran_hook[id], True
lock.release()
if prev:
    return
fn(*args, **kwargs)
```